File: 27-Trie/python/06_maximum_xor_of_two_numbers_in_an_array.py
from typing import List

class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        n = len(nums)
        trie = [[-1, -1] for _ in range(31 * n + 1)]
        trie_size = 1

        for num in nums:
            node = 0
            for k in range(30, -1, -1):
                bit = (num >> k) & 1
                if trie[node][bit] == -1:
                    trie[node][bit] = trie_size
                    trie_size += 1
                node = trie[node][bit]

        max_xor = 0
        for num in nums:
            node = 0
            current_xor = 0
            for k in range(30, -1, -1):
                bit = (num >> k) & 1
                opposite_bit = 1 - bit

                if trie[node][opposite_bit] != -1:
                    current_xor |= (1 << k)
                    node = trie[node][opposite_bit]
                else:
                    node = trie[node][bit]
            max_xor = max(max_xor, current_xor)

        return max_xor

File: 27-Trie/java/06-MaximumXorOfTwoNumbersInAnArray.java
class Solution {
    public int findMaximumXOR(int[] nums) {
        int[][] trie = new int[31 * nums.length + 1][2];
        int trieSize = 1;

        for (int[] node : trie) {
            node[0] = -1;
            node[1] = -1;
        }

        for (int num : nums) {
            int node = 0;
            for (int k = 30; k >= 0; k--) {
                int bit = (num >> k) & 1;
                if (trie[node][bit] == -1) {
                    trie[node][bit] = trieSize++;
                }
                node = trie[node][bit];
            }
        }

        int maxXor = 0;
        for (int num : nums) {
            int node = 0;
            int currentXor = 0;
            for (int k = 30; k >= 0; k--) {
                int bit = (num >> k) & 1;
                int oppositeBit = 1 - bit;

                if (trie[node][oppositeBit] != -1) {
                    currentXor |= (1 << k);
                    node = trie[node][oppositeBit];
                } else {
                    node = trie[node][bit];
                }
            }
            maxXor = Math.max(maxXor, currentXor);
        }

        return maxXor;
    }
}
